pkg/kube/adapter.go

```diff
@@ -0,0 +1,309 @@
+
+	// The type of Patch. Currently we only allow "JSONPatch".
+	PatchType *string `json:"patchType,omitempty"`
+
+	// AuditAnnotations is an unstructured key value map set by remote admission controller (e.g. error=image-blacklisted).
+	// MutatingAdmissionWebhook and ValidatingAdmissionWebhook admission controller will prefix the keys with
+	// admission webhook name (e.g. imagepolicy.example.com/error=image-blacklisted). AuditAnnotations will be provided by
+	// the admission webhook to add additional context to the audit log for this request.
+	AuditAnnotations map[string]string `json:"auditAnnotations,omitempty"`
+}
+
+func AdmissionReviewKubeToAdapter(object runtime.Object) (*AdmissionReview, error) {
+	var typeMeta metav1.TypeMeta
+	var req *AdmissionRequest
+	var resp *AdmissionResponse
+	switch obj := object.(type) {
+	case *kubeApiAdmissionv1beta1.AdmissionReview:
+		typeMeta = obj.TypeMeta
+		arv1beta1Response := obj.Response
+		arv1beta1Request := obj.Request
+		if arv1beta1Response != nil {
+			resp = &AdmissionResponse{
+				UID:     arv1beta1Response.UID,
+				Allowed: arv1beta1Response.Allowed,
+				Result:  arv1beta1Response.Result,
+				Patch:   arv1beta1Response.Patch,
+			}
+			if arv1beta1Response.PatchType != nil {
+				patchType := string(*arv1beta1Response.PatchType)
+				resp.PatchType = &patchType
+			}
+		}
+		if arv1beta1Request != nil {
+			req = &AdmissionRequest{
+				UID:       arv1beta1Request.UID,
+				Kind:      arv1beta1Request.Kind,
+				Resource:  arv1beta1Request.Resource,
+				UserInfo:  arv1beta1Request.UserInfo,
+				Name:      arv1beta1Request.Name,
+				Namespace: arv1beta1Request.Namespace,
+				Operation: string(arv1beta1Request.Operation),
+				Object:    arv1beta1Request.Object,
+				OldObject: arv1beta1Request.OldObject,
+			}
+		}
+
+	case *kubeApiAdmissionv1.AdmissionReview:
+		typeMeta = obj.TypeMeta
+		arv1Response := obj.Response
+		arv1Request := obj.Request
+		if arv1Response != nil {
+			resp = &AdmissionResponse{
+				UID:     arv1Response.UID,
+				Allowed: arv1Response.Allowed,
+				Result:  arv1Response.Result,
+				Patch:   arv1Response.Patch,
+			}
+			if arv1Response.PatchType != nil {
+				patchType := string(*arv1Response.PatchType)
+				resp.PatchType = &patchType
+			}
+		}
+
+		if arv1Request != nil {
+			req = &AdmissionRequest{
+				UID:       arv1Request.UID,
+				Kind:      arv1Request.Kind,
+				Resource:  arv1Request.Resource,
+				UserInfo:  arv1Request.UserInfo,
+				Name:      arv1Request.Name,
+				Namespace: arv1Request.Namespace,
+				Operation: string(arv1Request.Operation),
+				Object:    arv1Request.Object,
+				OldObject: arv1Request.OldObject,
+			}
+		}
+
+	default:
+		return nil, errors.Errorf("unsupported type :%v", object.GetObjectKind())
+	}
+
+	return &AdmissionReview{
+		TypeMeta: typeMeta,
+		Request:  req,
+		Response: resp,
+	}, nil
+}
+
+func AdmissionReviewAdapterToKube(ar *AdmissionReview, apiVersion string) runtime.Object {
```

Comment:
Should this also be added to validation webhook server as well?

Reply:
ohhh great question. I'll let @howardjohn or @elfinhe answer this :)

Reply:
> Should this also be added to validation webhook server as well?

I am not sure about this. What's jasonwzm@'s advice on this? Jason is an expert in k8s apis.

Reply:
I think it makes sense to add this to the validation server as well

Reply:
Thanks Tariq, since AdmissionReview is shared between both webhooks.

Reply:
great, thanks